### DISCUSSION TITLE
Improving coding standards

### DISCUSSION THREAD

**Discussion by ErikssonJoakim** [okp4/community#4] [OKP4 Portal]
This is a discussion to improve coding standards in the repository. We are numerous contributeurs and as a consequence the code we write looks and act differently from one another. This can create [smelly code](https://en.wikipedia.org/wiki/Code_smell) that result in long code review sessions and unnecessary coding mistakes.

To help us we have `ESLint` and `Prettier` to format and check problems in our code but tools can only take us so far. To improve our code further we also need to improve our coding output.

Following a meeting with @lolottetheclash and @JoeyDjr we have concluded the following items of discussion:

- standardizing `readonly | Readonly<T> | DeepReadonly<T>`. Right now we are using these types differently depending on the author and component. We need some common grounds on what to use and when. 

<details>
<summary><b>Examples</b></summary>

![Capture d’écran 2022-08-04 à 14 28 24](https://user-images.githubusercontent.com/75730728/182821992-eda5b9a8-6ba3-4704-b503-83f609dca0e9.png)

![Capture d’écran 2022-08-04 à 14 28 24](https://user-images.githubusercontent.com/75730728/182822813-4ff862ba-5bef-454b-97e4-14f575ece4f1.png)

![Capture d’écran 2022-08-04 à 14 28 24](https://user-images.githubusercontent.com/75730728/182823890-9e2c067e-3c19-4668-8489-b481cc5f25db.png)

![Capture d’écran 2022-08-04 à 14 28 24](https://user-images.githubusercontent.com/75730728/182824062-b19b9f3f-3b65-4a94-b1e2-dfb57b565c92.png)
</details>

---

- Do we pass props and if so when? As of now we have `BabylonEngine`, `Canvas` and `Typography` that passes their props to their children. 

<details>
<summary><b>Example</b></summary>

![Capture d’écran 2022-08-04 à 14 28 24](https://user-images.githubusercontent.com/75730728/182829886-cc06729f-f3a8-420c-9252-222903b577f9.png)
</details>

---

- Some import paths do not use the path aliases. We should use aliases as much as possible.

<details>
<summary><b>Examples</b></summary>

![Capture d’écran 2022-08-04 à 14 28 24](https://user-images.githubusercontent.com/75730728/182831521-1bb4a7a1-1b0a-40b5-8cf9-e8dffb6a1ffb.png)

![Capture d’écran 2022-08-04 à 14 28 24](https://user-images.githubusercontent.com/75730728/182831700-ca64b81e-ae9b-43ea-895c-f267a3a0350b.png)

![Capture d’écran 2022-08-04 à 14 28 24](https://user-images.githubusercontent.com/75730728/182833569-1f4d3564-c739-4155-9cb2-fa6516e685bd.png)
</details>

---

- On the subject of imports, should they be ordered and if so how? We think it would be a good idea.

---

- Should methods within components be obligated to use `useCallback` or `useMemo`, and do we demand that `useMemo` is used when possible? See #395
<details>
<summary><b>Example</b></summary>

![Capture d’écran 2022-08-04 à 14 28 24](https://user-images.githubusercontent.com/75730728/182832274-514771ab-bc79-43c4-ad27-942bfc398805.png)

![Capture d’écran 2022-08-04 à 14 28 24](https://user-images.githubusercontent.com/75730728/182834347-bb5c197a-70a7-4296-96a1-af4522a4ddca.png)
</details>

---

- Do we allow logical operations in the return of the component, and if so how many/ how complex? or should all logical operations be made before the return statement? 

<details>
<summary><b>Examples</b></summary>

![Capture d’écran 2022-08-04 à 14 28 24](https://user-images.githubusercontent.com/75730728/182835151-2aa795a8-dd4e-4196-8ed5-96af64be4351.png) 

![Capture d’écran 2022-08-04 à 14 28 24](https://user-images.githubusercontent.com/75730728/182835713-bcb2715a-7ea9-4434-b93d-fc360abe6f5f.png)

![Capture d’écran 2022-08-04 à 14 28 24](https://user-images.githubusercontent.com/75730728/182835919-5cfc17ec-c4f2-4a10-bb42-7cc6f32f0155.png)
</details>

---

- Should functions and methods be commented? If so when? Should they for example be commented only when they are exported, always or never? Comments on exported functions give a description when hovering the function which can be wanted by some.

<details>
<summary><b>Example</b></summary>

![Capture d’écran 2022-08-04 à 14 28 24](https://user-images.githubusercontent.com/75730728/182846994-f60ec5f0-cb02-4f71-8077-7156fc45ad17.png)
</details>

---

- Should we be commenting components?

<details>
<summary><b>Example</b></summary>

![Capture d’écran 2022-08-04 à 15 30 22](https://user-images.githubusercontent.com/75730728/182859271-4416375d-3f45-4d72-b8b5-cf53170a3e97.png)
</details>

---

- As discussed we should remove the T prefix from types.

<details>
<summary><b>Example</b></summary>

![Capture d’écran 2022-08-04 à 15 35 16](https://user-images.githubusercontent.com/75730728/182860350-723218b4-aa9e-4b76-8d43-1386b5edf736.png)
</details>

---

- How should the property types be styled? We propose to put the obligatory properties first and to end all property descriptions with correct punctuation.

<details>
<summary><b>Examples</b></summary>

![Capture d’écran 2022-08-04 à 15 37 54](https://user-images.githubusercontent.com/75730728/182860926-917a0199-06a5-43f8-9d87-032e194b63cc.png)

![Capture d’écran 2022-08-04 à 15 38 24](https://user-images.githubusercontent.com/75730728/182861033-0a5e7f32-1adb-449d-8fa1-f67fa68dc855.png)
</details>

---

- Should we be typing our components as [React.FC](https://medium.com/@harrymt/should-you-use-react-fc-for-typing-react-components-62cde9ba67c)? This is a divisive issue with no clear consensus 

<details>
<summary><b>Example</b></summary>
OBS this is from a PR not yet merged

![Capture d’écran 2022-08-04 à 15 44 59](https://user-images.githubusercontent.com/75730728/182862475-87cd1c30-cb6c-4b2d-af5b-be03ec113b4c.png)
</details>

---

- Should ...`-main` always be the first class name of a component? And Should components always hava a `main` class name in its first element tag even if the classNames is empty?

<details>
<summary><b>Examples</b></summary>

![Capture d’écran 2022-08-04 à 15 48 46](https://user-images.githubusercontent.com/75730728/182863363-c9f96890-1d82-4f92-a948-7e77c8a3d824.png)

![Capture d’écran 2022-08-04 à 15 52 00](https://user-images.githubusercontent.com/75730728/182864043-7445b23f-3337-434d-96d9-962e91ca2e99.png)
</details>

---

- Should class names strictly separate words with an hyphen or do we allow words put together?

<details>
<summary><b>Examples</b></summary>

![Capture d’écran 2022-08-04 à 15 55 49](https://user-images.githubusercontent.com/75730728/182864931-e8b39316-5bd2-4408-be0e-ad60bf354bbf.png)

![Capture d’écran 2022-08-04 à 15 56 15](https://user-images.githubusercontent.com/75730728/182865030-20f6f706-bcac-4b75-a855-1ba73c462065.png)
</details>

---

- We should remove unnecessary props that have the default value.

<details>
<summary><b>Examples</b></summary>

![Capture d’écran 2022-08-04 à 15 59 42](https://user-images.githubusercontent.com/75730728/182865826-86eba36b-fe77-41cc-8201-e9e784300ffb.png)

![Capture d’écran 2022-08-04 à 15 58 54](https://user-images.githubusercontent.com/75730728/182865650-24298eb2-88a7-4771-8a88-7ebedb6be5f5.png)
</details>

**Comment by ErikssonJoakim** [comment]
@ccamel @fredericvilcot

**Comment by lolottetheclash** [comment]
Concerning React.FC, there are several explanations for avoiding using it:

- it implies that the component can receive children, unless you type it as a void component
- it breaks the defaultProps (well, in principle it is not used in all cases)
- it does not keep the generic type 
and some other reasons...

You can have a look here: https://github.com/facebook/create-react-app/pull/8177 (there are several other articles that discuss the subject)

Maybe it's better to use JSX.Element with typing, what do you think?

**Reply by ErikssonJoakim** [reply]
I'm for removing React.FC. They way I see it we should be as explicit as possible in how we want our components to behave, and in that regard using React.FC is not preferable.

We could still by default give the option to pass children to our components if we want to, we only have to use the `React.PropsWitchChildren<props>` type.
![Capture d’écran 2022-08-05 à 12 26 26](https://user-images.githubusercontent.com/75730728/183058526-51aaf414-cddc-4a8e-8fd8-78622064cf30.png)

This will also give us the possibility to use generic properties, which I imagine is something we might want to do at some point.

**Reply by ccamel** [reply]
Agree. This is a general consensus today.

**Reply by ErikssonJoakim** [reply]
@fredericvilcot What do you think about removing React.FC from our components?

**Reply by fredericvilcot** [reply]
@ErikssonJoakim 
I fully agree with that, but in a second time ;)
Let's refactor the stories first 😌

**Comment by ErikssonJoakim** [comment]
> * standardizing `readonly | Readonly<T> | DeepReadonly<T>`. Right now we are using these types differently depending on the author and component. We need some common grounds on what to use and when.

Correct me if I'm wrong, but doesn't `DeepReadonly` recursively make all properties and their values `readonly`?

DeepReadonly on object --> DeepReadonlyObject, --> DeepReadonly on every `key` in object, --> DeepReadonlyMap on `key` --> DeepReadonly on `key` and `value` etc

<details>
<summary>DeepReadonly</summary>

![Capture d’écran 2022-08-05 à 13 27 56](https://user-images.githubusercontent.com/75730728/183068306-ebec9635-dd11-45ee-bff0-e7c0b2001de0.png)
</details>

If so it's enough to just use `DeepReadonly` as shown below, without any `readonly` modifiers.

```tsx
export type ComponentProps = DeepReadonly<{
/**
* property description.
*/
property: value
}>

export const Component: React.FC<ComponentProps> = ({property}: ComponentProps): JSX.Element => {/*code here... */}
```

**Reply by ccamel** [reply]
This is true. And I am in favor of removing the `readonly` keywords when the type is wrapped by `DeepReadonly`.

**Comment by JoeyDjr** [comment]
How about refactoring the `utils.ts` file by creating one utils file per type: `file-utils.ts`, `date-utils.ts`, `string-utils.ts`, `type-utils.ts`, etc.
@fredericvilcot @ErikssonJoakim @lolottetheclash  @ccamel

**Reply by ccamel** [reply]
@JoeyDjr IMHO, this issue should be addressed in another PR as it is structurally deeper and touches on the way functional/business segregation is defined. There is a lot to be said for this and that is why it should be addressed specifically.

**Comment by lolottetheclash** [comment]
It will also be necessary to rework the naming of boolean variables and properties: prefix them with `is` as in `isOpened` or `has` as in `hasError`, in order to be consistent with the naming.
